x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupFromIndexService.java

```diff
@@ -0,0 +1,144 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.enrich;
+
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.BlockStreamInput;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.authz.privilege.ClusterPrivilegeResolver;
+import org.elasticsearch.xpack.esql.action.EsqlQueryAction;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamOutput;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * {@link LookupFromIndexService} performs lookup against a Lookup index for
+ * a given input page. See {@link AbstractLookupService} for how it works
+ * where it refers to this process as a {@code LEFT JOIN}. Which is mostly is.
+ */
+public class LookupFromIndexService extends AbstractLookupService<LookupFromIndexService.Request, LookupFromIndexService.TransportRequest> {
+    public static final String LOOKUP_ACTION_NAME = EsqlQueryAction.NAME + "/lookup_from_index";
+
+    public LookupFromIndexService(
+        ClusterService clusterService,
+        SearchService searchService,
+        TransportService transportService,
+        BigArrays bigArrays,
+        BlockFactory blockFactory
+    ) {
+        super(
+            LOOKUP_ACTION_NAME,
+            ClusterPrivilegeResolver.MONITOR_ENRICH.name(), // TODO some other privilege
+            clusterService,
+            searchService,
+            transportService,
+            bigArrays,
+            blockFactory,
+            TransportRequest::readFrom
+        );
+    }
+
+    @Override
+    protected TransportRequest transportRequest(LookupFromIndexService.Request request, ShardId shardId) {
+        return new TransportRequest(
+            request.sessionId,
+            shardId,
+            request.inputDataType,
+            request.inputPage,
+            request.extractFields,
+            request.matchField
+        );
+    }
+
+    @Override
+    protected QueryList queryList(TransportRequest request, SearchExecutionContext context, Block inputBlock, DataType inputDataType) {
+        MappedFieldType fieldType = context.getFieldType(request.matchField);
+        return QueryList.termQueryList(fieldType, context, inputBlock, inputDataType);
```

Comment:
If the input data is multilvalued we'll match each value with this. Do we want it?

Reply:
We can sort out the semantics afterwards - for now, it's fine if the SV values match.
A few options:
a. match exactly the values (so [1,2,3] need a [1,2,3])
b. decompose the key similar to stats (so [1,2,3] becomes [1],[2],[3])
c. let the user pick by specifying an expression on the key

I'd favor option a) since it's less surprising. b and c can be emulated by the user by introducing an eval (which we would do in the planner, if we were to add this type of functionality).

Reply:
For now, I'm going to keep this as is - but we have the ability to change it as needed.